src/main/java/com/aws/greengrass/shadowmanager/ShadowManager.java

```diff
@@ -101,22 +104,34 @@ public class ShadowManager extends PluginService {
     private final CloudDataClient cloudDataClient;
     private final MqttClient mqttClient;
     private final PubSubIntegrator pubSubIntegrator;
+    private final ExecutorService executorService;
     private final AtomicReference<Strategy> currentStrategy = new AtomicReference<>(DEFAULT_STRATEGY);
+    // This is used from within the mqtt thread only
+    private Future<?> mqttCallbackFuture = new CompletableFuture<>();
     public final MqttClientConnectionEvents callbacks = new MqttClientConnectionEvents() {
         @Override
         public void onConnectionInterrupted(int errorCode) {
-            stopSyncingShadows(true);
+            handleAsync(() -> stopSyncingShadows(true));
         }
 
         @Override
         public void onConnectionResumed(boolean sessionPresent) {
-            if (inState(State.RUNNING)) {
-                startSyncingShadows(StartSyncInfo.builder().startSyncStrategy(true)
-                        .updateCloudSubscriptions(true).build());
+            handleAsync(() -> {
+                if (inState(State.RUNNING)) {
+                    startSyncingShadows(
+                            StartSyncInfo.builder().startSyncStrategy(true).updateCloudSubscriptions(true).build());
+                }
+            });
+        }
 
+        private void handleAsync(Runnable runnable) {
+            if (!mqttCallbackFuture.isDone() && !mqttCallbackFuture.isCancelled()) {
+                mqttCallbackFuture.cancel(true);
```

Comment:
is it a problem if mqtt disconnects and connects quickly - `stopSyncingShadows` is in-progress and cancelled prematurely before the next `startSyncingShadows`

Reply:
feels like maybe stopSyncingShadows shouldn't be async

Reply:
https://github.com/aws-greengrass/aws-greengrass-shadow-manager/blob/36bb2636d35b3dfd0719e12ab5d3f20562157056/src/main/java/com/aws/greengrass/shadowmanager/sync/strategy/BaseSyncStrategy.java#L195  there may be some waiting while stopping as it is synchronized. 

But, I'm updating this as it feels like we need to wait and there's not anything we're doing with the mqttclient itself.